website/content/en/preview/getting-started/getting-started-with-terraform/_index.md

```diff
@@ -185,55 +183,35 @@ using [IRSA](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/
 We will create the ServiceAccount and connect it to this role during the Helm
 chart install.
 
+Add the following to your `main.tf` to create the IAM role for the Karpenter service account.
+
 ```hcl
-module "iam_assumable_role_karpenter" {
-  source                        = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
-  version                       = "4.7.0"
-  create_role                   = true
-  role_name                     = "karpenter-controller-${var.cluster_name}"
-  provider_url                  = module.eks.cluster_oidc_issuer_url
-  oidc_fully_qualified_subjects = ["system:serviceaccount:karpenter:karpenter"]
-}
+module "karpenter_irsa" {
```

Comment:
We're still likely to have some problems with IAM permissions due to the condition in this [karpenter policy](https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks/policies.tf#L456-L459).

Reply:
could you elaborate on "likely to have some problems"? The wildcard permissions for run/terminate any EC2 instance without a scoping condition are going to be a tough hurdle to pass for security conscious/enterprise environments - restricting these permissions to only those resources with certain attributes (i.e. - tags/naming scheme) is the current standard practice in the community. You can see this in the other policies such as https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks/policies.tf#L699-L702

Is there a different tag or set of conditions we should be specifying?

Reply:
Sure, let me clarify.  

The [condition](https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks/policies.tf#L456-L459) I mentioned will match for the the `ec2:TerminateInstances` and `ec2:DeleteLaunchTemplates` actions as long as we also provide the same tag to the Karpenter Provisioner spec.  (By providing a tag in the Provisioner spec, you are telling Karpenter to apply said tag to the Launch Templates, Instances, and volumes it creates).
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  provider:
    tags:
        karpenter.sh/discovery: karpenter-demo
```

The problem I'm seeing is with the `ec2:RunInstances` action.  The IAM condition applied to the `RunInstances` action expects a tag to be on all the resources the RunInstances action is attempting to use.  But, if I understand correctly, this will never be able to match for some resources.  I've done some testing and [this IAM policy](https://gist.github.com/dewjam/93a7cdde5a170a1fefbd2b0182fc22b4) appears to work, though.

Reply:
Thanks for sharing @dewjam - I've opened a draft of a policy that I *hope will work based on the information you shared https://github.com/terraform-aws-modules/terraform-aws-iam/pull/209

However, how do I go about testing this change since `make test` and `make battletest` do not appear to fully exercise Karpenter? Do I just need to use this https://karpenter.sh/v0.7.2/getting-started/getting-started-with-terraform/#first-use and verify Karpenter is able to provision and delete nodes?

Reply:
> Do I just need to use this https://karpenter.sh/v0.7.2/getting-started/getting-started-with-terraform/#first-use and verify Karpenter is able to provision and delete nodes?

Yep exactly.

`make test` and `make battletest` spawn etcd and kube-apiserver as processes on your local machine.  This works well for testing Karpenter behaviors, but obviously doesn't take into account underlying cloud infrastructure like security groups and/or IAM policies.  We are working on a more robust e2e testing methodology, though.

Reply:
we should have https://github.com/terraform-aws-modules/terraform-aws-iam/pull/209 merged tomorrow and then this should be ready to re-check. testing locally though, Karpenter was able to provision and remove nodes (scale up/down) per the getting started guide. stay tuned